src/internal/kopia/base_finder.go

```diff
@@ -285,49 +272,125 @@ func (b *baseFinder) findBasesInSet(
 				"empty backup stream store ID",
 				"search_backup_id", bup.ID)
 
-			if !foundIncomplete {
-				foundIncomplete = true
+			continue
+		}
 
-				kopiaAssistSnaps = append(kopiaAssistSnaps, ManifestEntry{
+		// If we've made it to this point then we're considering the backup
+		// complete as it has both an item data snapshot and a backup details
+		// snapshot.
+		//
+		// Check first if this is an assist backup. Criteria for selecting an
+		// assist backup are:
+		// 1. most recent assist backup for that reason.
+		// 2. At most one assist backup per reason.
+		// 3. It must be more recent than the merge backup for the reason, if
+		// a merge backup exists.
+		if assistModel == nil {
+			assistModel = b.getAssistBackupModel(ictx, bup, reason)
+			if assistModel != nil {
+				assistSnap := ManifestEntry{
 					Manifest: man,
 					Reasons:  []Reasoner{reason},
-				})
+				}
+				kopiaAssistSnaps = append(kopiaAssistSnaps, assistSnap)
+
+				assistBackup = &backupBase{
+					BackupEntry:   assistModel,
+					ManifestEntry: &assistSnap,
+				}
 
 				logger.Ctx(ictx).Infow(
-					"found incomplete backup",
-					"search_backup_id", bup.ID)
-			}
+					"found assist backup",
+					"assist_backup_id", bup.ID,
+					"ssid", ssid)
 
-			continue
+				continue
+			}
 		}
 
-		// If we've made it to this point then we're considering the backup
-		// complete as it has both an item data snapshot and a backup details
-		// snapshot.
+		// Found the most recent merge backup. We can stop here.
+		//
+		// TODO(pandeyabs): Ideally we would do some sanity checks here, e.g.
+		// 1. Check for MergeBackup tag. Not safe, since we may have merge
+		// backups with older versions of corso which won't have the tag.
+		// 2. Check for ErrorCount == 0. We will enforce this after we remove
+		// support for BestEffort mode.
 		logger.Ctx(ictx).Infow("found complete backup", "base_backup_id", bup.ID)
 
-		me := ManifestEntry{
+		mergeSnap := ManifestEntry{
 			Manifest: man,
 			Reasons:  []Reasoner{reason},
 		}
-		kopiaAssistSnaps = append(kopiaAssistSnaps, me)
 
-		return &BackupEntry{
+		kopiaAssistSnaps = append(kopiaAssistSnaps, mergeSnap)
+
+		mergeModel := &BackupEntry{
 			Backup:  bup,
 			Reasons: []Reasoner{reason},
-		}, &me, kopiaAssistSnaps, nil
+		}
+
+		mergeBackup = &backupBase{
+			BackupEntry:   mergeModel,
+			ManifestEntry: &mergeSnap,
+		}
+
+		break
 	}
 
-	logger.Ctx(ctx).Info("no base backups for reason")
+	logger.Ctx(ctx).Info("no merge or assist backup for reason")
```

Comment:
nit: could you please update this log message? We'll reach here even if we do find valid bases